drivers/base/I2C_Device.zig

```diff
@@ -0,0 +1,112 @@
+    ///
+    /// Reserved addresses are ones that match `0b0000XXX` or `0b1111XXX`.
+    ///
+    /// See more here: https://www.i2c-bus.org/addressing/
+    pub fn is_reserved(addr: I2CAddress) bool {
+        const value: u7 = @intFromEnum(addr);
+        return ((value & 0x78) == 0) or ((value & 0x78) == 0x78);
+    }
+
+    pub fn format(addr: I2CAddress, fmt: []const u8, options: std.fmt.FormatOptions, writer: anytype) !void {
+        _ = fmt;
+        _ = options;
+        try writer.print("I2C(0x{X:0>2})", .{@intFromEnum(addr)});
+    }
+};
+
+// TODO: Can make this file the struct
+pub const I2C_Device = struct {
+    /// Pointer to the object implementing the driver.
+    ///
+    /// If the implementation requires no `ptr` pointer,
+    /// you can safely use `undefined` here.
+    ptr: *anyopaque,
+
+    /// Virtual table for the datagram device functions.
+    vtable: *const VTable,
+
+    pub fn set_address(dev: I2C_Device, addr: I2CAddress) I2CErrorI!void {
+        const set_address_fn = dev.vtable.set_address_fn orelse return I2CErrorI.Unsupported;
+        return set_address_fn(dev.ptr, addr);
+    }
+
+    /// Writes a single `datagram` to the device.
+    pub fn write(dev: I2C_Device, datagram: []const u8) I2CErrorI!void {
+        return try dev.writev(&.{datagram});
+    }
+
+    /// Writes multiple `datagrams` to the device.
+    pub fn writev(dev: I2C_Device, datagrams: []const []const u8) I2CErrorI!void {
+        const writev_fn = dev.vtable.writev_fn orelse return I2CErrorI.Unsupported;
+        return writev_fn(dev.ptr, datagrams);
+    }
+
+    /// Writes then reads a single `datagram` to the device.
+    pub fn write_then_read(dev: I2C_Device, src: []const u8, dst: []u8) I2CErrorI!void {
+        return try dev.writev_then_readv(&.{src}, &.{dst});
+    }
+
+    /// Writes a slice of datagrams to the device, then reads back into another slice of datagrams
+    pub fn writev_then_readv(
+        dev: I2C_Device,
+        write_chunks: []const []const u8,
+        read_chunks: []const []u8,
+    ) I2CErrorI!void {
+        const writev_then_readv_fn = dev.vtable.writev_then_readv_fn orelse return I2CErrorI.Unsupported;
+        return writev_then_readv_fn(dev.ptr, write_chunks, read_chunks);
+    }
+
+    /// Reads a single `datagram` from the device.
+    /// Function returns the number of bytes written in `datagram`.
+    pub fn read(dev: I2C_Device, datagram: []u8) I2CErrorI!usize {
+        return try dev.readv(&.{datagram});
+    }
+
+    /// Reads multiple `datagrams` from the device.
+    /// Function returns the number of bytes written in `datagrams`.
+    pub fn readv(dev: I2C_Device, datagrams: []const []u8) I2CErrorI!usize {
+        const readv_fn = dev.vtable.readv_fn orelse return I2CErrorI.Unsupported;
+        return readv_fn(dev.ptr, datagrams);
+    }
+
+    pub const VTable = struct {
+        set_address_fn: ?*const fn (*anyopaque, I2CAddress) I2CErrorI!void,
```

Comment:
Other than the error types being adjusted, this is the only change from the datagram interface.

Does it make sense to add anything else?
Does it make sense to remove any of these?